### PR TITLE
packaging: cache: Fix caching kernels which rely on extra modules

### PR DESCRIPTION
**Pull request by fidencio**
This logic has been broken forever, let's try to fix it here.

Please, see each commit message for more details and, please, do the best review possible on this one.

### CODE REVIEW

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 1]
```diff
@@ -126,6 +126,24 @@ EOF
...
 
+get_kernel_modules_dir() {
+	local kernel_version="${1:-}"
+	local kernel_kata_config_version="${2:-}"
+	local kernel_name"=${3:-}"
+	[ -z "${kernel_version}" ] && die "kernel version is a required argument"
+	[ -z "${kernel_kata_config_version}" ] && die "kernel kata config version is a required argument"
+	[ -z "${kernel_name}" ] && die "kernel name is a required argument"
+
+	local numeric_final_version=${kernel_version#v}
```
Comment (stevenhorsman) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:137]: This value isn't being used? Should it replace `${kernel_version#v}` in the echo, or be removed?

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 2]
```diff
@@ -966,6 +966,21 @@ handle_build() {
 	fi
 	tar tvf "${final_tarball_path}"
 
+	case ${build_target} in
+		kernel*-confidential|kernel-sev)
```
Comment (stevenhorsman) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:965]: Just to check, is the idea behind `kernel*-confidential` that we might have build_targets of kernel-confidential and kernel-<tee>-confidentials?

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 3]
```diff
@@ -372,22 +363,26 @@ install_kernel_helper() {
...
-		default_patches_dir="${repo_root_dir}/tools/packaging/kernel/patches"
-		module_dir="${repo_root_dir}/tools/packaging/kata-deploy/local-build/build/kernel-sev/builddir/kata-linux-${kernel_version#v}-${kernel_kata_config_version}/lib/modules/${kernel_version#v}"
 	elif [[ "${kernel_name}" == "kernel"*"-confidential" ]]; then
 		kernel_version="$(get_from_kata_deps assets.kernel.confidential.version)"
-		default_patches_dir="${repo_root_dir}/tools/packaging/kernel/patches"
-		module_dir="${repo_root_dir}/tools/packaging/kata-deploy/local-build/build/kernel-confidential/builddir/kata-linux-${kernel_version#v}-${kernel_kata_config_version}/lib/modules/${kernel_version#v}"
 	fi
 
-	install_cached_kernel_tarball_component ${kernel_name} ${module_dir} && return 0
+	if [ "${kernel_name}" == "kernel-sev" ] || [ "${kernel_name}" == "kernel"*"-confidential" ]; then
```
Comment (stevenhorsman) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:377]: does this need double braces to get the wildcard matching properly:
```
$ kernel_name=kernel-confidential; [ "${kernel_name}" == "kernel"*"-confidential" ] && echo "matched"
$ kernel_name=kernel-confidential; [[ "${kernel_name}" == "kernel"*"-confidential" ]] && echo "matched"
matched
```

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh (hunk 2 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:965]: Actually, I'd like to match `kernel-confidential`  and `kernel-nvidia-gpu-confidential`.

Comment (stevenhorsman) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:965]: Cool, that was just a (poor) example. Thanks for confirming

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh (hunk 1 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh]: Fixes, thanks!

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh (hunk 3 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh]: Fixed, thanks!

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 4]
```diff
@@ -966,6 +966,21 @@ handle_build() {
 	fi
 	tar tvf "${final_tarball_path}"
 
+	case ${build_target} in
+		kernel*-confidential|kernel-sev)
+			local modules_final_tarball_path="${workdir}/kata-static-${build_target}-modules.tar.xz"
+			if [ ! -f "${modules_final_tarball_path}" ]; then
+				local modules_dir=$(get_kernel_modules_dir ${kernel_version} ${kernel_kata_config_version})
```
Comment (beraldoleal) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:968]: A small suggestion is to squash this commit with the first one that you are introducing the function.

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 5]
```diff
@@ -996,7 +996,25 @@ handle_build() {
...
+				;;
+			*)
+				sudo oras push \
+					${ARTEFACT_REGISTRY}/kata-containers/cached-artefacts/${build_target}:latest-${TARGET_BRANCH}-$(uname -m) \
+					${final_tarball_name} \
+					${build_target}-version \
+					${build_target}-builder-image-version \
+					{build_target}-sha256sum
+				;;
+		esac
```
Comment (beraldoleal) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:1012]: Another suggestion would be  to try removing this duplication. Something like this (if possible):

```bash
additional_arg=""
case ${build_target} in
    kernel*-confidential|kernel-sev)
        additional_arg="kata-static-${build_target}-modules.tar.xz"
        ;;
esac

sudo oras push \
    ${ARTEFACT_REGISTRY}/kata-containers/cached-artefacts/${build_target}:latest-${TARGET_BRANCH}-$(uname -m) \
    ${final_tarball_name} \
    ${additional_arg:+$additional_arg} \
    ${build_target}-version \
    ${build_target}-builder-image-version \
    ${build_target}-sha256sum
```

I haven´t tested this code, so not sure if will work as it is.

**Review by beraldoleal — APPROVED** [review]
Hi @fidencio, I haven't tested the changes... So my LGTM is from the bash PoV. Take it with a  grant of salt.

Just left two small suggestions for you.

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 6]
```diff
@@ -141,6 +172,9 @@ install_cached_tarball_component() {
 	local current_image_version="${3}"
 	local component_tarball_name="${4}"
 	local component_tarball_path="${5}"
+	# extra_tarballs must be in the following format:
```
Comment (portersrc) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:175]: Looks like you future-proofed this format here? I saw one case where `extra_tarballs` gets set in your new changes (line 380), and it just specifies one tarball (not a space-separated list). Seems reasonable to allow for more in the future, but thought it worth asking.

**Review by portersrc — APPROVED** [review]
this looks good to me; nice work

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh (hunk 6 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:175]: I future proofed this as I can see NVIDIA wanting to also cache their deb files generated from the kernel build.

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh (hunk 5 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:1012]: I'll do this in a follow-up, just to make sure things work first with the "dumber" version.

**Pull request merged**